### PR TITLE
[14.0][IMP] product_barcode_constraint_per_company: simplify logic

### PR DESCRIPTION
**Pull request by dessanhemrayev**
Change old method
Update oca_dependencies file
Fix tests

### CODE REVIEW

File: product_barcode_constraint_per_company/__manifest__.py
Comment (legalsylvain) [inline product_barcode_constraint_per_company/__manifest__.py:11]: could you preserve authorship, and add Cetmix instead of replacing ? thanks !

Comment (dessanhemrayev) [reply product_barcode_constraint_per_company/__manifest__.py:11]: okay, thanks

**Review by legalsylvain — APPROVED** [review]

File: product_barcode_constraint_per_company/README.rst
Comment (legalsylvain) [inline product_barcode_constraint_per_company/README.rst]: No need to remove this file. It générate useless diff.

File: product_barcode_constraint_per_company/__manifest__.py
Comment (legalsylvain) [inline product_barcode_constraint_per_company/__manifest__.py]: This new dependency is more problematic.

**Review by legalsylvain — CHANGES_REQUESTED** [review]
Do not add new dependency to multi barcode module.

**Comment by dessanhemrayev** [comment]
> Do not add new dependency to multi barcode module.

@legalsylvain   Do you think we need to create a bridge module for that, or is there possibly another solution?

**Comment by legalsylvain** [comment]
multi barcode module is a module that is really changing a lot of things. The feature is not desired in some implementation. So in that case, indeed glue module looks a good solution, if you want to make the module product_barcode_constraint_per_company working with that module.

**Comment by francesco-ooops** [comment]
supersedes #526

**Review by legalsylvain — COMMENTED** [review]
except the deletion of the README.rst file, LGTM. Thanks !

**Review by legalsylvain — COMMENTED** [review]
nitpicking : still changes in product_barcode_constraint_per_company/README.rst ! 

No need to change this file, it will be generated automatically, when mergin the PR, by the bot.

regards.

File: product_barcode_constraint_per_company/tests/test_module.py
Comment (geomer198) [inline product_barcode_constraint_per_company/tests/test_module.py]: Please add assertion message for all assert cases and docstring for all tests functions

File: product_multi_barcode_constraint_per_company/models/product_barcode.py
Comment (geomer198) [inline product_multi_barcode_constraint_per_company/models/product_barcode.py]: Why you use search function here?
And why you call `super` in cycle?

File: product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py
Comment (geomer198) [inline product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py]: You deactivate parent function. Why do you use `super` here?

File: product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py
Comment (geomer198) [inline product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py]: You deactivate parent function. Why do you use `super` here?

**Review by geomer198 — CHANGES_REQUESTED** [review]
Please fix comments

File: product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py
Comment (dessanhemrayev) [reply product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py]: Thanks, I'll fix it

File: product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py
Comment (dessanhemrayev) [reply product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py]: Thanks, I'll fix it

File: product_multi_barcode_constraint_per_company/models/product_barcode.py
Comment (dessanhemrayev) [reply product_multi_barcode_constraint_per_company/models/product_barcode.py]: Thanks, I'll fix it

File: product_multi_barcode_constraint_per_company/models/product_barcode.py
Comment (geomer198) [inline product_multi_barcode_constraint_per_company/models/product_barcode.py:13]: For what you write `search` here?

File: product_multi_barcode_constraint_per_company/models/product_barcode.py
Comment (geomer198) [inline product_multi_barcode_constraint_per_company/models/product_barcode.py]: Cycle into cycle it's bad solution, please leave only one cycle.

File: product_multi_barcode_constraint_per_company/models/product_barcode.py
Comment (geomer198) [inline product_multi_barcode_constraint_per_company/models/product_barcode.py]: Why do you removed `super`?

File: product_multi_barcode_constraint_per_company/models/product_barcode.py
Comment (geomer198) [inline product_multi_barcode_constraint_per_company/models/product_barcode.py]: `dict()` slowly than `{}`, so better use `{}` in code.

**Review by geomer198 — CHANGES_REQUESTED** [review]
@dessanhemrayev Please, fix my comments.

File: product_multi_barcode_constraint_per_company/models/product_barcode.py
Comment (dessanhemrayev) [reply product_multi_barcode_constraint_per_company/models/product_barcode.py]: okay, thanks

File: product_multi_barcode_constraint_per_company/models/product_barcode.py
Comment (dessanhemrayev) [reply product_multi_barcode_constraint_per_company/models/product_barcode.py]: okay, I'll fix it

File: product_multi_barcode_constraint_per_company/models/product_barcode.py
Comment (dessanhemrayev) [reply product_multi_barcode_constraint_per_company/models/product_barcode.py]: okay, I'll fix it

File: product_multi_barcode_constraint_per_company/models/product_barcode.py
Comment (geomer198) [inline product_multi_barcode_constraint_per_company/models/product_barcode.py]: Please change like this:
```py
@api.constrains("name")
    def _check_duplicates(self):
        existing_codes = self.env["product.barcode"]
        for record in self:
            barcode = self.search(
                [("id", "!=", record.id), ("name", "=", record.name)], limit=1
            )
            if (
                barcode
                and barcode.sudo().product_tmpl_id.company_id.id
                == record.product_tmpl_id.company_id.id
            ):
                existing_codes |= record
                break
        if existing_codes:
            super(ProductBarcode, existing_codes)._check_duplicates()
```

**Review by geomer198 — CHANGES_REQUESTED** [review]
Please fix my comment

**Review by geomer198 — APPROVED** [review]
LGTM!

**Comment by geomer198** [comment]
@legalsylvain Could you please check changes?

**Review by legalsylvain — APPROVED** [review]
Partial review . (product_barcode_constraint_per_company only)
LGTM. Thanks for work !

**Comment by francesco-ooops** [comment]
@florentx good for you?

**Comment by francesco-ooops** [comment]
@OCA/logistics-maintainers good for merge?

File: product_multi_barcode_constraint_per_company/__init__.py
Comment (simahawk) [inline product_multi_barcode_constraint_per_company/__init__.py]: what is this?

File: product_barcode_constraint_per_company/tests/test_module.py
Comment (simahawk) [inline product_barcode_constraint_per_company/tests/test_module.py]: there's no init file here, these tests won't run at all

File: product_barcode_constraint_per_company/tests/test_module.py
Comment (simahawk) [inline product_barcode_constraint_per_company/tests/test_module.py]: 2023 comes after :wink:

File: product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py
Comment (simahawk) [inline product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py]: I don't get what is this void thing

**Review by simahawk — CHANGES_REQUESTED** [review]

File: oca_dependencies.txt
Comment (simahawk) [inline oca_dependencies.txt]: FTR this file is useless for OCA builds since a while (unless you rely on it for your CI)

File: product_multi_barcode_constraint_per_company/__init__.py
Comment (dessanhemrayev) [reply product_multi_barcode_constraint_per_company/__init__.py]: to skip tests that don't work correctly without a new module

File: product_barcode_constraint_per_company/tests/test_module.py
Comment (dessanhemrayev) [reply product_barcode_constraint_per_company/tests/test_module.py]: I don't understand you, what did you mean here?

File: oca_dependencies.txt
Comment (dessanhemrayev) [reply oca_dependencies.txt]: Thanks, I'll fix it

File: product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py
Comment (dessanhemrayev) [reply product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py]: Codecov requires that this void be covered with a test

File: product_barcode_constraint_per_company/tests/test_module.py
Comment (dessanhemrayev) [reply product_barcode_constraint_per_company/tests/test_module.py]: Please check carefully, the init file is in the repository. It is not displayed because this file is immutable
image

**Comment by francesco-ooops** [comment]
@simahawk better now?

File: product_barcode_constraint_per_company/tests/test_module.py
Comment (simahawk) [reply product_barcode_constraint_per_company/tests/test_module.py]: You should not change the order. Leave the original (c) line on top

File: product_barcode_constraint_per_company/tests/test_module.py
Comment (simahawk) [reply product_barcode_constraint_per_company/tests/test_module.py]: sorry, I mixed w/ the other init file in this PR :v:

File: product_multi_barcode_constraint_per_company/__init__.py
Comment (simahawk) [reply product_multi_barcode_constraint_per_company/__init__.py]: I still don't understand why you need this. Can you please elaborate and explain why it is needed?
BTW you should not import here any test stuff.

**Review by simahawk — CHANGES_REQUESTED** [review]

File: product_multi_barcode_constraint_per_company/__init__.py
Comment (dessanhemrayev) [reply product_multi_barcode_constraint_per_company/__init__.py]: Without disabling the tests, there is no point in checking the functionality of the modules. 
I disabled the tests of the module `product_barcode_constraint_per_company` because its tests do not work correctly after we changed the module itself, that is, we moved the company_id field to a separate module, and changed the verification method. now, after we have made changes, we run the tests, they crash, the tests are due to the module `product_multi_barcode`. The module `product_multi_barcode` does not provide for the situation that one barcode can be in different companies. Because of this, we had to write a new bridge module. In which all these situations are fixed.

File: product_barcode_constraint_per_company/tests/test_module.py
Comment (dessanhemrayev) [reply product_barcode_constraint_per_company/tests/test_module.py]: Thanks, I fix it)

File: product_multi_barcode_constraint_per_company/__init__.py
Comment (francesco-ooops) [reply product_multi_barcode_constraint_per_company/__init__.py]: @simahawk makes sense now?

**Comment by francesco-ooops** [comment]
@simahawk good to go?

File: product_barcode_constraint_per_company/tests/test_module.py
Comment (HekkiMelody) [inline product_barcode_constraint_per_company/tests/test_module.py]: **todo**: it's not necessary to tag these tests as `post_install`. Please remove

File: product_barcode_constraint_per_company/tests/test_module.py
Comment (HekkiMelody) [inline product_barcode_constraint_per_company/tests/test_module.py]: **todo**: Please restore this to its original value

File: product_multi_barcode_constraint_per_company/models/product_barcode.py
Comment (HekkiMelody) [inline product_multi_barcode_constraint_per_company/models/product_barcode.py]: **question**: By default `product.barcode` doesn't have a `company_id`, therefore it's visible to everyone from every company. Does it cause issues if, for example, people can see multiple of the same barcode, but can't necessarily see the related products?

Should barcode also have a `company_id` field (maybe a related field, from `product_tmpl_id.company_id`?)

In the case of the constraint, I think it may be better to just override it. If `company_id` is False, there can be no other barcode with the same name. If `company_id` has a value, it has to be unique per company.  Please add a test for the case where `company_id` is False.

File: product_multi_barcode_constraint_per_company/README.rst
Comment (HekkiMelody) [inline product_multi_barcode_constraint_per_company/README.rst]: **todo**: Pre-commit should have regenerated this README (and also the `index.html`) automatically, I'm not sure why it didn't do that. Can you check?

File: product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py
Comment (HekkiMelody) [reply product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py]: I agree with @simahawk , I don't think this is appropriate or necessary

File: product_multi_barcode_constraint_per_company/__init__.py
Comment (HekkiMelody) [reply product_multi_barcode_constraint_per_company/__init__.py]: I don't think this is necessary.

In OCA tests are run only once, on module installation. If your new module breaks the old module's tests, that shouldn't be a problem. It's normal for your module to change the behaviour of a related module, but the tests are executed in installation order.

Everything in this file (except the first line) should be removed IMO.

File: product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py
Comment (HekkiMelody) [inline product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py]: If you want to inherit the setup from your other tests, but not the tests themselves, you should separate the setup function into its own class so you can inherit from it separately.

Here's an example: https://github.com/OCA/stock-logistics-workflow/blob/14.0/stock_picking_auto_create_lot/tests/test_stock_picking_auto_create_lot.py#L10

**Review by HekkiMelody — CHANGES_REQUESTED** [review]
This PR is definitely not very clean. I think the general concept is sound, but there are some issues to fix.

File: product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py
Comment (HekkiMelody) [inline product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py]: **issue**: These tests should use the new `product.barcode` model more extensively. This includes checking a product that has two barcodes, for example.

File: product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py
Comment (dessanhemrayev) [reply product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py]: Thanks for the advice, I will definitely use it.

File: product_multi_barcode_constraint_per_company/README.rst
Comment (dessanhemrayev) [reply product_multi_barcode_constraint_per_company/README.rst]: It may be necessary to update the pre-commit

File: product_barcode_constraint_per_company/tests/test_module.py
Comment (dessanhemrayev) [reply product_barcode_constraint_per_company/tests/test_module.py]: ok, I'll fix it

File: product_barcode_constraint_per_company/tests/test_module.py
Comment (dessanhemrayev) [reply product_barcode_constraint_per_company/tests/test_module.py]: ok, I'll fix it

File: product_barcode_constraint_per_company/__manifest__.py
Comment (dessanhemrayev) [reply product_barcode_constraint_per_company/__manifest__.py]: ok, I'll fix it)

File: product_multi_barcode_constraint_per_company/__init__.py
Comment (dessanhemrayev) [reply product_multi_barcode_constraint_per_company/__init__.py]: okay, I'll fix it

File: product_multi_barcode_constraint_per_company/models/product_barcode.py
Comment (dessanhemrayev) [reply product_multi_barcode_constraint_per_company/models/product_barcode.py]: Thanks for the advice, I will definitely use it.

File: product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py
Comment (dessanhemrayev) [reply product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py]: okay,  I will definitely add a test for it.

File: product_barcode_constraint_per_company/tests/test_module.py
Comment (HekkiMelody) [inline product_barcode_constraint_per_company/tests/test_module.py]: **issue**: I noticed that tests broke because of the order in which modules are installed and tested in the OCA CI.

1. product_multi_barcode
2. product_barcode_constraint_per_company --> tests break here
3. product_multi_barcode_constraint_per_company

The tests break, predictably, because the glue module hasn't been installed yet. Because of this, I think in this specific circumstance it may be acceptable to tag these tests as `post_install`, with a comment explaining why it is being done this way.

File: product_multi_barcode_constraint_per_company/models/product_barcode.py
Comment (HekkiMelody) [inline product_multi_barcode_constraint_per_company/models/product_barcode.py]: **chore**: you don't need to access the `id` here. Also, just to really make sure that the order of operations is clear and unambiguous, I think I'd prefer more brackets

```suggestion
                barcode
                and (not record.product_id.company_id or
                    (barcode.sudo().product_id.company_id == record.product_id.company_id)
                )
            ):
```

File: product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py
Comment (HekkiMelody) [inline product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py]: **chore**: this is the default; IMO this line is not necessary

File: product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py
Comment (HekkiMelody) [inline product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py]: (non-blocking)

In case you didn't know, you can decorate the entire test method with `mute_logger("odoo.sql_db")` so you don't have to specify it every time here.

**Review by HekkiMelody — CHANGES_REQUESTED** [review]
Code review, just a couple of minor issues IMO. Much better than before.

**question(non-blocking)**: not sure about the use of newsfragments in OCA anymore. In this case, what would `2944` refer to?

File: product_barcode_constraint_per_company/tests/test_module.py
Comment (dessanhemrayev) [reply product_barcode_constraint_per_company/tests/test_module.py]: Thank you so much for your help, this advice helped me fix this problem with the tests.

File: product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py
Comment (dessanhemrayev) [reply product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py]: Thanks, I'll fix it

File: product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py
Comment (dessanhemrayev) [reply product_multi_barcode_constraint_per_company/tests/test_product_barcode_constraint.py]: I'll fix it

**Review by HekkiMelody — APPROVED** [review]
Code review, LGTM

**Comment by francesco-ooops** [comment]
@simahawk good for you?

**Comment by francesco-ooops** [comment]
@simahawk waiting for your blessing :)

**Comment by francesco-ooops** [comment]
@simahawk help me Orsi-Wan Kenobi, you're my only hope

**Comment by francesco-ooops** [comment]
@pedrobaeza can this be merged based on existing reviews?

File: product_barcode_constraint_per_company/README.rst
Comment (pedrobaeza) [inline product_barcode_constraint_per_company/README.rst]: Remove this change.

File: product_barcode_constraint_per_company/__manifest__.py
Comment (pedrobaeza) [inline product_barcode_constraint_per_company/__manifest__.py]: Don't change this. It was correct. Now, it isn't, as the spaces are included.

File: product_barcode_constraint_per_company/__manifest__.py
Comment (pedrobaeza) [inline product_barcode_constraint_per_company/__manifest__.py:15]: You mustn't change dependencies in a stable module.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: product_barcode_constraint_per_company/__manifest__.py
Comment (francesco-ooops) [reply product_barcode_constraint_per_company/__manifest__.py:15]: Hi @pedrobaeza , AFAICS this is not a stable module, and the change in dependency was asked here: https://github.com/OCA/stock-logistics-barcode/pull/526#pullrequestreview-1658253604 following https://github.com/OCA/stock-logistics-barcode/pull/526#discussion_r1275978467

Comment (pedrobaeza) [reply product_barcode_constraint_per_company/__manifest__.py:15]: OK then about this.

File: product_barcode_constraint_per_company/__manifest__.py
Comment (simahawk) [inline product_barcode_constraint_per_company/__manifest__.py]: Detail: as you added a new module in the same PR you should change the version manually here, unless you don't care that the new module will have version 1.1.0 too when we'll bump w/ `minor`.

File: product_multi_barcode_constraint_per_company/models/product_barcode.py
Comment (simahawk) [inline product_multi_barcode_constraint_per_company/models/product_barcode.py]: I'd move `sudo` to `for record in self.sudo()`

**Review by simahawk — COMMENTED** [review]
pls check Pedro's remarks

File: product_barcode_constraint_per_company/__manifest__.py
Comment (dessanhemrayev) [reply product_barcode_constraint_per_company/__manifest__.py]: Thanks for review!)
I'll fix it

File: product_multi_barcode_constraint_per_company/models/product_barcode.py
Comment (dessanhemrayev) [reply product_multi_barcode_constraint_per_company/models/product_barcode.py]: Thanks for review!)
I'll fix it

**Comment by francesco-ooops** [comment]
@pedrobaeza @simahawk kind reminder :)

File: product_barcode_constraint_per_company/README.rst
Comment (simahawk) [reply product_barcode_constraint_per_company/README.rst]: still here

**Review by simahawk — APPROVED** [review]
minor remark, pre-approving

**Comment by dessanhemrayev** [comment]
@pedrobaeza  please check, I have corrected your comments)
Thanks

**Review by pedrobaeza — APPROVED** [review]
/ocabot merge nobump

**Pull request merged**
